cmake/Platform/Linux/CompilerSettings_linux.cmake

```diff
@@ -6,29 +6,30 @@
 #
 #
 
-if(NOT CMAKE_C_COMPILER AND NOT CMAKE_CXX_COMPILER AND NOT "$ENV{CC}" AND NOT "$ENV{CXX}")
-    set(path_search
-        /bin
-        /usr/bin
-        /usr/local/bin
-        /sbin
-        /usr/sbin
-        /usr/local/sbin
+function(find_default_clang)
+    set(supported_clang_version_numbers 12 11 10 9 8 7 6.0)
+    set(supported_clang_versions ${supported_clang_version_numbers})
+    list(TRANSFORM supported_clang_versions PREPEND "clang-")
+    set(supported_clangxx_versions ${supported_clang_version_numbers})
+    list(TRANSFORM supported_clangxx_versions PREPEND "clang++-")
+
+    find_program(
+        CMAKE_C_COMPILER
+        NAMES
+            $ENV{CC}
+            clang
+            ${supported_clang_versions}
+        REQUIRED
+    )
+
+    find_program(
+        CMAKE_CXX_COMPILER
+        NAMES
+            $ENV{CXX}
+            clang++
+            ${supported_clangxx_versions}
+        REQUIRED
     )
-    list(TRANSFORM path_search APPEND "/clang-[0-9]*")
-    file(GLOB clang_versions ${path_search})
-    if(clang_versions)
-        # Find and pick the highest installed version
-        list(SORT clang_versions COMPARE NATURAL)
```

Comment:
The previous algorithm was not been used if `CMAKE_C_COMPILER`, CMAKE_CXX_COMPILER`, etc were defined. That should stay. The definition of those variables point to the actual versions that the user wants. If that binary is not there, the new version will silently pick up something else. 

The previous algorithm didn't require to have a hardcoded list of clang versions, this will require us to continuously update the list. 

The reason why is not picking the higher version seems to be in this sort. In https://cmake.org/cmake/help/latest/command/list.html#ordering it states:
> NATURAL: Sorts a list of strings using natural order (see strverscmp(3) manual), i.e. such that contiguous digits are compared as whole numbers. For example: the following list 10.0 1.1 2.1 8.0 2.0 3.1 will be sorted as 1.1 2.0 2.1 3.1 8.0 10.0 if the NATURAL comparison is selected where it will be sorted as 1.1 10.0 2.0 2.1 3.1 8.0 with the STRING comparison.

So, line 23 should pick the last one instead of the first one.

This also needs to be updated in the project templates

Reply:
`find_program` does not redo the search if the cache variable it would use to store the result of the search is already set. So, if the user does `cmake ... -DCMAKE_C_COMPILER=gcc`, the call to `find_program(CMAKE_C_COMPILER NAMES clang)` does not execute any search. However, I did find that `find_program` will mark the output cache variable's type as a `FILEPATH`, which, according to the `set()` documentation will:
> if the <type> is PATH or FILEPATH and the <value> provided on the command line is a relative path, then the set command will treat the path as relative to the current working directory and convert it to an absolute path.

This really stinks, because it forces the user to specify an absolute path to their compiler, when that's not required without the `find_program` call.

Instead of doing any of that, I found that we can simply set the `CMAKE_C_COMPILER_NAMES` and `CMAKE_CXX_COMPILER_NAMES` variables to the appropriate list, and that will control CMake's built-in default compiler search.

Reply:
from my previous tests with `COMPILER_NAMES`, I found that the code is still picking gcc before considering `COMPILER_NAMES`. Take a look at `Modules\CMakeDetermineCompiler.cmake` in CMake (macro `_cmake_find_compiler`)
Note that:
```
    foreach(l ${_languages})
      list(APPEND CMAKE_${lang}_COMPILER_LIST
        ${_${lang}_COMPILER_NAMES_${CMAKE_${l}_COMPILER_ID}})
    endforeach()
    # Prefer vendors based on the platform.
    list(APPEND CMAKE_${lang}_COMPILER_LIST ${CMAKE_${lang}_COMPILER_NAMES})
```
That means that there will be a compiler list with other compilers before `CMAKE_<lang>_COMPILER_NAMES` is added

We should also not call those variables `supported_`, we dont support all those versions.

Reply:
Changing this code to use `find_package` or `CMAKE_<lang>_COMPILER_NAMES` seems like more trouble than it is worth, if we want to detect which versioned clang executable the user has installed. So I just reverted all of that, and changed the sort to say `ORDER DESCENDING` instead.

Reply:
sounds good, the final remaining thing is to update the template files